Lib/fontTools/subset/svg.py

```diff
@@ -0,0 +1,235 @@
+from __future__ import annotations
+
+import re
+from itertools import chain, count, groupby
+from typing import Dict, Iterable, Iterator, List, Set, Tuple
+
+try:
+    from lxml import etree
+except ModuleNotFoundError:
+    # lxml is required for subsetting SVG, but we prefer to delay the import error
+    # until subset_glyphs() is called (i.e. if font to subset has an 'SVG ' table)
+    etree = None
+
+from fontTools import ttLib
+from fontTools.subset.util import _add_method
+
+
+__all__ = ["subset_glyphs"]
+
+
+GID_RE = re.compile(r"^glyph(\d+)$")
+
+NAMESPACES = {
+    "svg": "http://www.w3.org/2000/svg",
+    "xlink": "http://www.w3.org/1999/xlink",
+}
+XLINK_HREF = f'{{{NAMESPACES["xlink"]}}}href'
+
+if etree is not None:
+    # it's more efficient to compile XPath objects upfront than calling the same
+    # Element.xpath() several times
+    find_svg_elements_with_id = etree.XPath(".//svg:*[@id]", namespaces=NAMESPACES)
+
+    # We currently support xlink:href (as used by <use> and gradient templates),
+    # and local url(#...) links found in fill or clip-path attributes
+    # TODO(anthrotype): Check we aren't missing other supported kinds of reference
+    find_svg_elements_with_references = etree.XPath(
+        ".//svg:*[ "
+        "starts-with(@xlink:href, '#') "
+        "or starts-with(@fill, 'url(#') "
+        "or starts-with(@clip-path, 'url(#') "
+        "or contains(@style, ':url(#') "
+        "]",
+        namespaces=NAMESPACES,
+    )
+
+    find_svg_elements_with_glyph_href = etree.XPath(
+        ".//svg:*[starts-with(@xlink:href, '#glyph')]", namespaces=NAMESPACES
+    )
+
+
+def group_elements_by_id(tree: etree.Element) -> Dict[str, etree.Element]:
+    return {el.attrib["id"]: el for el in find_svg_elements_with_id(tree)}
+
+
+def parse_css_declarations(style_attr: str) -> Dict[str, str]:
+    # https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/style
+    # https://developer.mozilla.org/en-US/docs/Web/CSS/Syntax#css_declarations
+    result = {}
+    for declaration in style_attr.split(";"):
+        if declaration.count(":") == 1:
+            property_name, value = declaration.split(":")
+            property_name = property_name.strip()
+            result[property_name] = value.strip()
+        elif declaration.strip():
+            raise ValueError(f"Invalid CSS declaration syntax: {declaration}")
+    return result
+
+
+def iter_referenced_ids(tree: etree.Element) -> Iterator[str]:
```

Comment:
Shouldn't the return type be Generator?

Reply:
the [python docs](https://docs.python.org/3.6/library/typing.html#typing.Generator) recommend to use the simpler Iterator[...] annotation when one doesn't use the Generator's SendType and ReturnType

Reply:
Oh nice, I think we have the wrong type all over nanoemoji then :D